aptos-move/framework/aptos-framework/sources/staking_contract.move

```diff
@@ -0,0 +1,1132 @@
+    ) acquires Store {
+        let staker_address = signer::address_of(staker);
+        assert_staking_contract_exists(staker_address, old_operator);
+
+        // Merging two existing staking contracts is too complex as we'd need to merge two separate stake pools.
+        let store = borrow_global_mut<Store>(staker_address);
+        let staking_contracts = &mut store.staking_contracts;
+        assert!(
+            !simple_map::contains_key(staking_contracts, &new_operator),
+            error::invalid_state(ECANT_MERGE_STAKING_CONTRACTS),
+        );
+
+        let (_, staking_contract) = simple_map::remove(staking_contracts, &old_operator);
+        // Force distribution of any already inactive stake.
+        distribute_internal(staker_address, old_operator, &mut staking_contract, &mut store.distribute_events);
+
+        // For simplicity, we request commission to be paid out first. This avoids having to ensure to staker doesn't
+        // withdraw into the commission portion.
+        request_commission_internal(
+            old_operator,
+            &mut staking_contract,
+            &mut store.add_distribution_events,
+            &mut store.request_commission_events,
+        );
+
+        // Update the staking contract's commission rate and stake pool's operator.
+        stake::set_operator_with_cap(&staking_contract.owner_cap, new_operator);
+        staking_contract.commission_percentage = new_commission_percentage;
+
+        let pool_address = staking_contract.pool_address;
+        simple_map::add(staking_contracts, new_operator, staking_contract);
+        emit_event(
+            &mut store.switch_operator_events,
+            SwitchOperatorEvent { pool_address, old_operator, new_operator }
+        );
+    }
+
+    /// Allow anyone to distribute already unlocked funds. This does not affect reward compounding and therefore does
+    /// not need to be restricted to just the staker or operator.
+    public entry fun distribute(staker: address, operator: address) acquires Store {
+        assert_staking_contract_exists(staker, operator);
+        let store = borrow_global_mut<Store>(staker);
+        let staking_contract = simple_map::borrow_mut(&mut store.staking_contracts, &operator);
+        distribute_internal(staker, operator, staking_contract, &mut store.distribute_events);
+    }
+
+    /// Distribute all unlocked (inactive) funds according to distribution shares.
+    fun distribute_internal(
+        staker: address,
+        operator: address,
+        staking_contract: &mut StakingContract,
+        distribute_events: &mut EventHandle<DistributeEvent>,
+    ) {
+        let pool_address = staking_contract.pool_address;
+        let (_, inactive, _, pending_inactive) = stake::get_stake(pool_address);
+        let total_potential_withdrawable = inactive + pending_inactive;
+        let coins = stake::withdraw_with_cap(&staking_contract.owner_cap, total_potential_withdrawable);
```

Comment:
what happened here? this is off... this will fail if pending_inactive > 0.

Reply:
Test for this edge case: https://github.com/aptos-labs/aptos-core/pull/3570/files#diff-8d7b802c58c375726f1ab691afdec810cf833b5f2d1d64536c7d3e6ebd97e179R1032